src/serprog.c

```diff
@@ -0,0 +1,510 @@
+#define S_CMD_Q_IFACE           0x01    // Query interface version
+#define S_CMD_Q_CMDMAP          0x02    // Query supported commands bitmap
+#define S_CMD_Q_PGMNAME         0x03    // Query programmer name
+#define S_CMD_Q_SERBUF          0x04    // Query Serial Buffer Size
+#define S_CMD_Q_BUSTYPE         0x05    // Query supported bustypes
+#define S_CMD_Q_CHIPSIZE        0x06    // Query supported chipsize (2^n format)
+#define S_CMD_Q_OPBUF           0x07    // Query operation buffer size
+#define S_CMD_Q_WRNMAXLEN       0x08    // Query Write to opbuf: Write-N maximum length
+#define S_CMD_R_BYTE            0x09    // Read a single byte
+#define S_CMD_R_NBYTES          0x0A    // Read n bytes
+#define S_CMD_O_INIT            0x0B    // Initialize operation buffer
+#define S_CMD_O_WRITEB          0x0C    // Write opbuf: Write byte with address
+#define S_CMD_O_WRITEN          0x0D    // Write to opbuf: Write-N
+#define S_CMD_O_DELAY           0x0E    // Write opbuf: udelay
+#define S_CMD_O_EXEC            0x0F    // Execute operation buffer
+#define S_CMD_SYNCNOP           0x10    // Special no-operation that returns NAK+ACK
+#define S_CMD_Q_RDNMAXLEN       0x11    // Query read-n maximum length
+#define S_CMD_S_BUSTYPE         0x12    // Set used bustype(s).
+#define S_CMD_O_SPIOP           0x13    // Perform SPI operation.
+#define S_CMD_S_SPI_FREQ        0x14    // Set SPI clock frequency
+#define S_CMD_S_PIN_STATE       0x15    // Enable/disable output drivers
+#define S_CMD_S_SPI_CS          0x16    // Set SPI chip select to use
+#define S_CMD_S_SPI_MODE        0x17    // Sets the spi mode used by S_CMD_O_SPIOP
+#define S_CMD_S_CS_MODE         0x18    // Sets the way the CS is controlled
+
+enum spi_mode {
+    SPI_MODE_HALF_DUPLEX = 0,
+    SPI_MODE_FULL_DUPLEX = 1,
+    SPI_MODE_MAX = SPI_MODE_FULL_DUPLEX,
+};
+
+enum cs_mode {
+    CS_MODE_AUTO = 0,
+    CS_MODE_SELECTED = 1,
+    CS_MODE_DESELECTED = 2,
+    CS_MODE_MAX = CS_MODE_DESELECTED,
+};
+
+// little endian helper functions
+
+static uint16_t read_le16(const unsigned char *buf) {
+    return buf[0] | (buf[1] << 8);
+}
+
+static uint32_t read_le32(const unsigned char *buf) {
+    return buf[0] | (buf[1] << 8) | (buf[2] << 16) | (buf[3] << 24);
+}
+
+static void write_le24(unsigned char *buf, uint32_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+    buf[2] = val >> 16;
+}
+
+static void write_le32(unsigned char *buf, uint32_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+    buf[2] = val >> 16;
+    buf[3] = val >> 24;
+}
+
+// serprog communication functions
+
+static int perform_serprog_cmd_full(const PROGRAMMER *pgm, uint8_t cmd,
+                                    const unsigned char *params, int params_len,
+                                    const unsigned char *send_buf, int send_len,
+                                    unsigned char *recv_buf, int recv_len) {
+    unsigned char resp_status_code = 0;
+
+    serial_send(&pgm->fd, &cmd, 1);
```

Comment:
It is good practice to check *every* function that can return errors whether it performed OK. This helps debugging if things go wrong, as you catch where a problem first surfaced. There are several techniques, eg, the use of `EI()` and `EV()` macros in `buspirate.c` or the pedestrian way of
```
   if(serial_send(&pgm->fd, &cmd, 1) < 0)
     return -1;
```